pennylane/optimize/rotosolve.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Rotosolve gradient free optimizer"""
+
+from pennylane import numpy as np
+
+
+class RotosolveOptimizer:
+    r"""Rotosolve gradient free optimizer.
+
+    The Rotosolve optimizer minimizes an objective function with respect to the parameters of a
+    quantum circuit without the need of calculating the gradient of the function. The algorithm
+    works by updating the parameters :math:`\theta = \theta_1, \dots, \theta_D` one at a time
+    according to a closed-form expression for the optimal value of the :math:`d^{th}` parameter
+    :math:`\theta^*_d` when the other parameters are fixed:
+
+    .. math:: \theta^*_d = \underset{\theta_d}{\text{argmin}}\left<H\right>_{\theta_d}
+              = -\frac{\pi}{2} - \text{arctan2}\left(2\left<H\right>_{\theta_d=0}
+              - \left<H\right>_{\theta_d=\pi/2} - \left<H\right>_{\theta_d=-\pi/2},
+              \left<H\right>_{\theta_d=\pi/2} - \left<H\right>_{\theta_d=-\pi/2}\right),
+
+    where :math:`\left<H\right>_{\theta_d}` is the expectation value of the objective function
+    optimized over the parameter :math:`\theta_d`. :math:`\text{arctan2}(x, y)` computes the
+    element-wise arc tangent of :math:`x/y` choosing the quadrant correctly, avoiding, in
+    particular, division-by-zero when :math:`y = 0`.
+
+    The algorithm is described in further detail in `Ostaszewski et al. (2019) <https://arxiv.org/abs/1905.09692>`_
+    """
+    # pylint: disable=too-few-public-methods
+
+    def __init__(self):
+        pass
+
+    def step(self, objective_fn, x):
+        """Update x with one step of the optimizer.
+
+        Args:
+            objective_fn (function): the objective function for optimization
+            x (array): NumPy array containing the current values of the variables to be updated
+
+        Returns:
+            array: the new variable values :math:`x^{(t+1)}`
+        """
+        # make sure that x is an array
+        if np.ndim(x) == 0:
+            x = np.array([x])
+
+        for d, _ in enumerate(x):
+            x = self._rotosolve(objective_fn, x, d)
```

Comment:
From my understanding, Rotosolve is a simple optimization strategy, whereas Rotoselect allows the structure of the circuit (selection of the generators) to be learnt as well. In this case, it doesn't make sense for Rotosolve to have the same restriction on the cost function signature (`cost(x, generator)`), and instead here `cost(x)` should simply be allowed. Is this taken into account with the class inheritance?

Reply:
I completely agree; and that should be the case in the code. In the ``Rotosolve._rotosolve`` method the ``objective_fn`` (same as ``cost``) only takes one argument (the list ``x``) and outputs a single value (the cost). It differs from the ``Rotoselect._rotosolve`` method where the ``objective_fn`` takes two arguments (the lists ``x`` and ``generators``). That method still only does the Rotosolve part of the Rotoselect optimization, hence the name. This is why I actually _didn't_ let Rotoselect inherit from Rotosolve, since all functions differ, and it was easier simply to rewrite the ``_rotosolve`` method for both than to either have cases for different signatures or to change the signature for ``objective_fn`` inside the function (doable, but more complicated than simply having two versions of the method).

The ``x`` and ``d`` arguments are needed in both cases, since ``x`` is the list of values and ``d`` is the position to update next. Does everything make sense? :slightly_smiling_face:

Reply:
This sounds reasonable :) Perhaps a `Rotoselect._rotosolve` -> `Rotoselect._rotoselect` or comparable renaming could help out with avoiding future confusions

Reply:
> This is why I actually didn't let Rotoselect inherit from Rotosolve, since all functions differ, and it was easier simply to rewrite the _rotosolve method for both than to either have cases for different signatures or to change the signature for objective_fn inside the function (doable, but more complicated than simply having two versions of the method).

Makes sense! Thanks @thisac